test/integration/protocol_integration_test.cc

```diff
@@ -4693,6 +4693,239 @@ TEST_P(ProtocolIntegrationTest, InvalidResponseHeaderNameStreamError) {
   ASSERT_TRUE(fake_upstream_connection_->connected());
 }
 
+// Validate that when allow_multiplexed_upstream_half_close is enabled a request with H/2 or H/3
+// upstream is not reset when upstream half closes before downstream and allows downstream to send
+// data even if upstream is half closed.
+// H/1 upstream always causes the stream to be closed if it responds before downstream.
+// This test also causes downstream connection to run out of stream window (in H/2 and H/3 case)
+// when processing END_STREAM from the server to make sure the data is not lost in the case.
+TEST_P(ProtocolIntegrationTest, ServerHalfCloseBeforeClientWithBufferedResponseData) {
```

Comment:
might be useful to dump the upstream timing and/or warn about it as I think before if we'd get an early response onLastUpstreamTxByte might not be set.

While now it will be and it might be > onLastUpstreamRxByteReceived which might cause some manipulations of timing data that assumed last byte tx to upstream < last upstream rx byte received to become negative

Reply:
I've added test for duration values. `%COMMON_DURATION%` may now produce negative values in more cases but I'm not too concerned about it, since independent half close support is basically a new feature. It is useful for bidi streaming in gRPC and I think people will make sure to set monitoring correctly.

Reply:
since independent half close support is basically a new feature. It is useful for bidi streaming in gRPC and I think people will make sure to set monitoring correctly.
> It seems like our plan is to eventually enable it by default. In which case warning or noting it in either release notes or the docs for `%COMMON_DURATION%` will help less sophisticated users understand when this can be negative.